mac/Keyman4MacIM/Keyman4MacIM/KMPackageReader.m

```diff
@@ -0,0 +1,337 @@
+  NSMutableArray *files = [NSMutableArray arrayWithCapacity:0];
+  NSMutableArray *keyboards = [NSMutableArray arrayWithCapacity:0];
+  KMPackageInfoBuilder *builder = [[KMPackageInfoBuilder alloc] init];
+
+  @try {
+    NSString *fileContents = [[NSString stringWithContentsOfFile:path encoding:NSWindowsCP1252StringEncoding error:NULL] stringByReplacingOccurrencesOfString:@"\r" withString:@""];
+    NSArray *lines = [fileContents componentsSeparatedByString:@"\n"];
+    ContentType contentType = ctUnknown;
+    for (NSString *line in lines) {
+      if (!line.length)
+        continue;
+
+      if ([[line lowercaseString] hasPrefix:[kPackage lowercaseString]]) {
+        contentType = ctPackage;
+        continue;
+      }
+      else if ([[line lowercaseString] hasPrefix:[kButtons lowercaseString]]) {
+        contentType = ctButtons;
+        continue;
+      }
+      else if ([[line lowercaseString] hasPrefix:[kStartMenu lowercaseString]]) {
+        contentType = ctStartMenu;
+        continue;
+      }
+      else if ([[line lowercaseString] hasPrefix:[kStartMenuEntries lowercaseString]]) {
+        contentType = ctStartMenuEntries;
+        continue;
+      }
+      else if ([[line lowercaseString] hasPrefix:[kInfo lowercaseString]]) {
+        contentType = ctInfo;
+        continue;
+      }
+      else if ([[line lowercaseString] hasPrefix:[kFiles lowercaseString]]) {
+        contentType = ctFiles;
+        continue;
+      }
+
+      switch (contentType) {
+        case ctPackage: {
+          if ([[line lowercaseString] hasPrefix:[kReadMeFile lowercaseString]])
+            builder.readmeFilename = [line substringFromIndex:kReadMeFile.length+1];
+          else if ([[line lowercaseString] hasPrefix:[kGraphicFile lowercaseString]])
+            builder.graphicFilename = [line substringFromIndex:kGraphicFile.length+1];
+          break;
+        }
+        case ctButtons:
+          break;
+        case ctStartMenu:
+          break;
+        case ctStartMenuEntries: {
+          if ([[line lowercaseString] hasPrefix:[kWelcome lowercaseString]]) {
+            NSString *s = [line substringFromIndex:kWelcome.length+1];
+            NSArray *vs = [s componentsSeparatedByString:@"\","];
+            NSString *v1 = [[vs objectAtIndex:0] stringByReplacingOccurrencesOfString:@"\"" withString:@""];
+            NSString *v2 = [[vs objectAtIndex:1] stringByReplacingOccurrencesOfString:@"\"" withString:@""];
+          }
+
+          break;
+        }
+        case ctInfo: {
+          if ([[line lowercaseString] hasPrefix:[kName lowercaseString]]) {
+            NSString *s = [line substringFromIndex:kName.length+1];
+            NSArray *vs = [s componentsSeparatedByString:@"\","];
+            NSString *v1 = [[vs objectAtIndex:0] stringByReplacingOccurrencesOfString:@"\"" withString:@""];
+            NSString *v2 = [[vs objectAtIndex:1] stringByReplacingOccurrencesOfString:@"\"" withString:@""];
+            builder.packageName = v1;
```

Comment:
This method for parsing the inf file is way too long and contains a lot of duplication, but I didn't really want to take on refactoring it. I mainly replaced the dictionary references with calls to the builder object and moved it out of the AppDelegate to this new class. I feel like if I were going to refactor this properly, then I would create two separate parser classes for each format but also delve into the things in the inf file that don't seem to be represented by the json file and figure out whether we need to load them. 

Not sure if the time is justified now because maybe the .inf format is going away? Or will it be around forever because some keyman packages have only the .inf file. Could an intern create an inf to json converter and update all the old packages that have an .inf only?__

Reply:
Yes, this might be a product of my jetlag. I didn't realise when I reviewed the first time that this was old code, just moved.

.inf will be around 'forever'. There are many packages out there that we don't develop or distribute so we can't change them. But there is an "if it ain't broke" reality and we can certainly spend our time better elsewhere.

So yep, good call, leave it as it is :grin: